bayesian/utility/user_utils.py

```diff
@@ -22,42 +26,59 @@ def get_user(uuid_val):
+    """Create or Update User."""
     try:
-        insert_user_stmt = insert(UserDetails).values(user_id=uuid_val, user_source=user_source,
-                                                      created_date=datetime.datetime.now())
-
-        rdb.session.execute(insert_user_stmt)
-        rdb.session.commit()
-        logger.info(f"User created with id {uuid_val}")
+        if snyk_api_token:
+            insert_user_stmt = insert(UserDetails).values(user_id=user_id, user_source=user_source,
+                                                          snyk_api_token=snyk_api_token,
+                                                          created_date=datetime.datetime.now(),
+                                                          status=UserStatus.REGISTERED.name,
+                                                          registered_date=datetime.datetime.now())
+
+            do_update_stmt = insert_user_stmt.on_conflict_do_update(
+                index_elements=['user_id'], set_=dict(snyk_api_token=snyk_api_token,
+                                                      updated_date=datetime.datetime.now()))
+            rdb.session.execute(do_update_stmt)
+            rdb.session.commit()
+            logger.info("User added with id %s", user_id)
     except SQLAlchemyError as e:
         rdb.session.rollback()
-        logger.exception("Error creating new user")
-        raise UserException("Error creating user") from e
+        logger.exception("Error updating user with id %s", user_id)
+        raise UserException("Error updating user") from e
 
 
-@retry(reraise=True, stop=tenacity.stop_after_attempt(3), wait=tenacity.wait_fixed(1))
-def update_user(user_id, snyk_api_token):
-    """Update User."""
+@retry(reraise=True, stop=tenacity.stop_after_attempt(6),
+       wait=tenacity.wait_exponential(multiplier=2, min=4))
+def is_snyk_token_valid(snyk_api_token):
+    """Validate Snyk API token."""
     try:
-        if snyk_api_token:
-            update_stmt = update(UserDetails).where(UserDetails.user_id == user_id).values(
-                dict(snyk_api_token=snyk_api_token, updated_date=datetime.datetime.now(),
-                     status=UserStatus.REGISTERED.name, registered_date=datetime.datetime.now()))
-
-        rdb.session.execute(update_stmt)
-        rdb.session.commit()
-        logger.info(f"User updated with id {user_id}")
-    except SQLAlchemyError as e:
-        rdb.session.rollback()
-        logger.exception(f"Error updating user with id {user_id}")
-        raise UserException("Error updating user") from e
+        response = requests.post(SNYK_API_TOKEN_VALIDATION_URL, json={'api': snyk_api_token},
+                                 headers={'Content-Type': 'application/json'})
+        if response.status_code == 200:
+            return True
+        else:
+            return False
+    except Exception as e:
+        logger.exception("Encountered exception calling Snyk")
+        raise e
```

Comment:
How about using Exception for flow control? it is quite common in Python world.
The given token is valid if there no Exeption, otherwise invalid.

```suggestion
response = requests.post(SNYK_API_TOKEN_VALIDATION_URL, json={'api': snyk_api_token})
response.raise_for_status()
```

Reply:
I am not sure I understand this, response.raise_for_status is fine for cases where the request is made. What about the case when the request failed, I would like to print out a log saying the request has failed.